homeassistant/components/knx/cover.py

```diff
@@ -38,21 +41,57 @@ async def async_setup_platform(
     discovery_info: DiscoveryInfoType | None = None,
 ) -> None:
     """Set up cover(s) for KNX platform."""
+    _async_migrate_unique_id(hass, discovery_info)
     entities = []
     for device in hass.data[DOMAIN].xknx.devices:
         if isinstance(device, XknxCover):
             entities.append(KNXCover(device))
     async_add_entities(entities)
 
 
+@callback
+def _async_migrate_unique_id(
+    hass: HomeAssistant, discovery_info: DiscoveryInfoType | None
+) -> None:
+    """Change unique_ids used in 2021.4 to include position_target GA."""
+    entity_registry = er.async_get(hass)
+    if not discovery_info or not discovery_info["platform_config"]:
+        return
+
+    platform_config = discovery_info["platform_config"]
+    for entity_config in platform_config:
+        # normalize group address strings - ga_updown was the old uid but is optional
+        updown_addresses = entity_config.get(CoverSchema.CONF_MOVE_LONG_ADDRESS)
+        if updown_addresses is None:
+            continue
+        ga_updown = parse_device_group_address(updown_addresses[0])
+        old_uid = str(ga_updown)
+
+        entity_id = entity_registry.async_get_entity_id("cover", DOMAIN, old_uid)
+        if entity_id is None:
+            continue
+        position_target_addresses = entity_config.get(
+            CoverSchema.CONF_POSITION_ADDRESS
+        )
+        ga_position_target = (
+            parse_device_group_address(position_target_addresses[0])
+            if position_target_addresses is not None
+            else None
+        )
+        new_uid = f"{ga_updown}_{ga_position_target}"
```

Comment:
If `ga_position_target` is `None` it seems weird to include it as a string `"None"` in the unique id.

Reply:
It's one of `move_long_address` or `position_address` or both. Including the "None" string was easier than forming a conditional string. 
I could change this in the evening (but I guess we'd miss Beta then)

Reply:
```suggestion
        if ga_updown is None and ga_position_target is None:
            # This should never happen.
            raise RuntimeError("Invalid cover address.")
        if ga_updown is None:
            new_uid = str(ga_position_target)
        elif ga_position_target is None:
            new_uid = str(ga_updown)
        else:
            new_uid = f"{ga_updown}_{ga_position_target}"
```

Reply:
Nevermind, I think we can keep it as just one case. The benefit with that is that it always has the same structure and we can map back to the components that make up the unique id.